File: 2022/OpenCV4 tutorial/morphology/morphology.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
using namespace cv;



int main( int argc, char** argv )
{
    auto x = MORPH_BLACKHAT;
    struct Userdata{
        Mat src;
        int morph_elem;
        int morph_size;
        int morph_operator;
    } userdata{.src = imread( "binary-icon-30.bmp", IMREAD_GRAYSCALE ),
               .morph_elem = 0,
               .morph_size=0,
               .morph_operator = 0} ;
    

    namedWindow( "Morphology Transformations Demo"); // Create window
    createTrackbar("Operator", "Morphology Transformations Demo", &userdata.morph_operator, 7, [](int Operator, void* userdata)->void{
        
        auto &ud = *static_cast<Userdata*>(userdata);
        ud.morph_operator = Operator;
        static Mat dst;
        Mat element = getStructuringElement( ud.morph_elem, Size( 2*ud.morph_size + 1, 2*ud.morph_size+1 ), Point( ud.morph_size, ud.morph_size ) );
        morphologyEx( ud.src, dst, ud.morph_operator, element );
        imshow( "Morphology Transformations Demo", dst );


    }, &userdata); //创建滑杆

    createTrackbar( "Element", "Morphology Transformations Demo", &userdata.morph_elem, 2, [](int elem_T, void* userdata)->void{
        auto &ud = *static_cast<Userdata*>(userdata);
        ud.morph_elem = elem_T;
        static Mat dst;
        Mat element = getStructuringElement( ud.morph_elem, Size( 2*ud.morph_size + 1, 2*ud.morph_size+1 ), Point( ud.morph_size, ud.morph_size ) );
        morphologyEx( ud.src, dst, ud.morph_operator, element );
        imshow( "Morphology Transformations Demo", dst );      
                        
    }, &userdata);
    createTrackbar( "Kernel size:(2n+1)", "Morphology Transformations Demo", &userdata.morph_size, 21, [](int morph_size, void* userdata)->void{
        auto &ud = *static_cast<Userdata*>(userdata);
        ud.morph_size = morph_size;
        static Mat dst;
        Mat element = getStructuringElement( ud.morph_elem, Size( 2*ud.morph_size + 1, 2*ud.morph_size+1 ), Point( ud.morph_size, ud.morph_size ) );
        morphologyEx( ud.src, dst, ud.morph_operator, element );
        imshow( "Morphology Transformations Demo", dst );   

    }, &userdata); //创建滑杆

    waitKey(0);
    return 0;
}

File: 2022/OpenCV4 tutorial/morphology/morphology.py
import cv2 as cv

def main():
    src = cv.imread("binary-icon-30.bmp", cv.IMREAD_GRAYSCALE)
    class Params:
        morph_elem = 0
        morph_size = 0
        morph_operator = 0
    def Operator_onChange(Operator):
        Params.morph_operator = Operator

        element = cv.getStructuringElement(Params.morph_elem, [2 * Params.morph_size + 1, 2 * Params.morph_size + 1], [Params.morph_size, Params.morph_size])
        dst = cv.morphologyEx(src, Params.morph_operator, element)
        cv.imshow("Morphology Transformations Demo", dst)

    cv.namedWindow("Morphology Transformations Demo") # Create window
    cv.createTrackbar("Operator", "Morphology Transformations Demo", Params.morph_operator, 7, Operator_onChange)

    def Element_onChange(elem_T):

        Params.morph_elem = elem_T
        element = cv.getStructuringElement(Params.morph_elem, [2 * Params.morph_size + 1, 2 * Params.morph_size + 1], [Params.morph_size, Params.morph_size])
        dst = cv.morphologyEx(src, Params.morph_operator, element)
        cv.imshow("Morphology Transformations Demo", dst)

    cv.createTrackbar("Element", "Morphology Transformations Demo", Params.morph_elem, 2, Element_onChange)

    def ksize_onChange(morph_s):

        Params.morph_size = morph_s
        element = cv.getStructuringElement(Params.morph_elem, [2 * Params.morph_size + 1, 2 * Params.morph_size + 1], [Params.morph_size, Params.morph_size])
        dst = cv.morphologyEx(src, Params.morph_operator, element)
        cv.imshow("Morphology Transformations Demo", dst)

    cv.createTrackbar("Kernel size:(2n+1)", "Morphology Transformations Demo", Params.morph_size, 21, ksize_onChange)

    cv.waitKey()


if __name__ == '__main__':
    main()
